(closes #91) Update Openmp GPU and add explicit memory versions to ACC

sergisiso opened the pull request.

sergisiso commented:
Ok, the picture that I am getting is that for this benchmark kernels is slightly faster than explicit loops and explicit loops do not improve the kernels in the mixed version. Also excplicit data movement makes the kernels slighlty faster but not the explicit loops. OpenMP is behind OpenACC, even with the version that looks the same.
![image](https://user-images.githubusercontent.com/1449555/208886610-e3b601e6-bb11-4da4-8d8f-086780004ef0.png)


These are all very small differences, if compared with CPU they are all much faster.
![image](https://user-images.githubusercontent.com/1449555/208886652-3e7b47a9-930f-474e-8803-265ddc8380d1.png)

sergisiso commented:
@arporter This is ready to review. In addition to the tracer_advection updates I also introduced a NemoLite2D OpenMP offloading script. It still doesn't work (like the ACC equivalent?) but the psyclone part and the compilation does and it is convenient to have it repo to to try it in different machines/compilers.

arporter commented on benchmarks/nemo/tracer_advection/scripts/acc_loops_explicit_data_movement_trans.py:
```diff
@@ -0,0 +1,78 @@
+
+    :param psy: the PSy object which this script will transform.
+    :type psy: :py:class:`psyclone.psyGen.PSy`
+    :returns: the transformed PSy object.
```
Blank line :returns: please.

sergisiso replied:
Done

arporter commented on benchmarks/nemo/tracer_advection/scripts/acc_loops_explicit_data_movement_trans.py:
```diff
@@ -0,0 +1,78 @@
+    for invoke in psy.invokes.invoke_list:
+        print(invoke.name)
+
+        normalise_loops(
```
Please add a comment to briefly say what this does.

sergisiso replied:
Done

arporter commented on benchmarks/nemo/tracer_advection/scripts/acc_loops_explicit_data_movement_trans.py:
```diff
@@ -0,0 +1,78 @@
+# Authors: S. Siso, STFC Daresbury Lab
+
+''' PSyclone transformation script to insert OpenACC Parallel Loop directives
+to the outermost loop that is parallelisable, including implicit loops.'''
```
Please also mention the explicit data-movement directives.

sergisiso replied:
Done

arporter commented on benchmarks/nemo/tracer_advection/scripts/acc_loops_explicit_data_movement_trans.py:
```diff
@@ -0,0 +1,78 @@
+        print(invoke.name)
+
+        normalise_loops(
+                invoke.schedule,
```
pylint complains about the indentation here.

sergisiso replied:
Fixed

arporter commented on benchmarks/nemo/tracer_advection/scripts/acc_mixed_explicit_data_movement_trans.py:
```diff
@@ -0,0 +1,207 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2018-2021, Science and Technology Facilities Council.
```
Please update the date.

sergisiso replied:
Done

arporter commented on benchmarks/nemo/tracer_advection/scripts/acc_mixed_explicit_data_movement_trans.py:
```diff
@@ -0,0 +1,207 @@
+
+'''
+
+from __future__ import print_function
```
Please rm this line.

sergisiso replied:
Done

arporter commented on benchmarks/nemo/tracer_advection/scripts/acc_mixed_explicit_data_movement_trans.py:
```diff
@@ -0,0 +1,207 @@
+            for assignment in sched.walk(Assignment):
+                ARRAY_RANGE_TRANS.apply(assignment)
+
+        add_kernels(sched.children)
```
Could/should use the add_kernels from utils.py?

sergisiso replied:
Yes.
Initially I didn't do it because the "loop independent" was added at the bottom of this call chain. But now I add the kernels and the loop directives independently in two steps and always using the utils infrastructure. The new version is slightly better because it manages to put more loop directives and collapse to 3 when possible.

arporter commented on benchmarks/nemo/tracer_advection/scripts/acc_mixed_explicit_data_movement_trans.py:
```diff
@@ -0,0 +1,207 @@
+Fortran. Note that the Fortran source files provided to PSyclone must
+have already been preprocessed (if required).
+
+The transformation script attempts to insert Kernels directives at the
```
If it's possible to re-write this script so as to use `utils.py` then please remove this chunk of the docstring.

sergisiso replied:
Done.

arporter commented on benchmarks/nemo/tracer_advection/scripts/acc_mixed_explicit_data_movement_trans.py:
```diff
@@ -0,0 +1,207 @@
+            continue
+        sched.view()
+
+        if EXPAND_IMPLICIT_LOOPS:
```
I'm a bit confused here. I thought 'mixed' meant a combination of the loop transformations with the use of ACC kernels but I can't see that this script does any loop transformations? (`COLLAPSE_LOOPS` and `EXPAND_IMPLICIT_LOOPS` are both `False`.)

sergisiso replied:
Yeah sorry, this was a remain of code before I decided that for benchmarking it would be easier to use calling different scripts instead of parameters. I removed all this now and always use the utils functions.

arporter commented on benchmarks/nemo/tracer_advection/scripts/omp_gpu_trans.py:
```diff
@@ -36,8 +36,7 @@
+from psyclone.psyir.transformations import OMPTargetTrans, OMPLoopTrans
 from utils import insert_explicit_loop_parallelism, normalise_loops
 
 
```
pylint complains a lot about the indentation in this file!

sergisiso replied:
I fixed them now.

arporter replied:
Weird. I've just tried again and it gets 10/10 now :-)

arporter commented on benchmarks/nemo/tracer_advection/scripts/utils.py:
```diff
@@ -120,3 +121,68 @@ def insert_explicit_loop_parallelism(
                 loop.parent.parent.collapse = num_nested_loops
+
+
+def valid_kernel(node):
```
This could be used in both acc_mixed_*_trans.py scripts.

sergisiso replied:
Using this function (and the insert_explicit_loop) from mixed_trans now.

arporter reviewed (changes requested):
Thanks for taking this on Sergi, it's a big job rationalising everything. I think we can punt the problems with NEMOLite to another issue and focus on tracer advection here.
I'm a bit confused about the 'mixed' transformation scripts - they don't seem to do what I thought they would. Apart from that, there's also some duplication of functionality between some of the scripts and `utils.py` that I think we can eliminate now?

sergisiso commented on benchmarks/nemo/nemolite2d/psykal/Makefile:
```diff
@@ -84,6 +84,20 @@ nemolite2d_omp: serial_libs nemolite2d_alg.f90 psyclone_scripts/omp_transform.py
 	cd $@ && make nemolite2d
 	@echo "OpenMP version prepared."
 
+# OpenMP GPU offloading PSyclone code-generation of NemoLite2D
```
I added the "with UM" and not working notices. I also suspect the issue is in the derived types.

sergisiso commented on benchmarks/nemo/nemolite2d/psykal/README.md:
```diff
@@ -32,10 +32,10 @@ The Makefile supports many targets:
 - nemolite2d_mpi - MPI version for distributed memory parallelism
 - nemolite2d_hybrid - MPI and OpenMP combined version
-- nemolite2d_acc - OpenACC (Not working - PSyclone will not generate it)
+- nemolite2d_acc - OpenACC offloading (Not working - issue with GPU routine with structures?)
```
Ups I copy pasted to much here. For me it also is the checksums returning 0.

sergisiso commented on benchmarks/nemo/nemolite2d/psykal/README.md:
```diff
@@ -32,10 +32,10 @@ The Makefile supports many targets:
 - nemolite2d_hybrid - MPI and OpenMP combined version
-- nemolite2d_acc - OpenACC (Not working - PSyclone will not generate it)
+- nemolite2d_acc - OpenACC offloading (Not working - issue with GPU routine with structures?)
+- nemolite2d_omp_gpu - OpenMP offloading (Not working - issue with GPU routine with structures?)
```
I will move this to another branch for now and if we don't find the cause I will create a minimal example of the issue.

sergisiso commented on benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_gpu_transform.py:
```diff
@@ -0,0 +1,49 @@
+''' Python script intended to be passed to PSyclone's generate()
+function via the -s option. It applies OpenMP to every loop and
+inlines all kernels in the schedule.'''
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_gpu_transform.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def trans(psy):
+    ''' Transformation entry point '''
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_gpu_transform.py:
```diff
@@ -0,0 +1,49 @@
+    loop_trans.omp_directive = "teamsdistributeparalleldo"
+    loop_trans.omp_schedule = "none"
+
+    for outer_loop in schedule.walk(Loop, stop_type=Loop):
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_gpu_transform.py:
```diff
@@ -0,0 +1,49 @@
+        except TransformationError:
+            pass
+
+    for kernel_subroutine in schedule.parent.walk(Routine):
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/serial_transform.py:
```diff
@@ -1,14 +1,13 @@
 
 
 def trans(psy):
     ''' Transformation script entry function '''
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/serial_transform.py:
```diff
@@ -1,14 +1,13 @@
 ''' Python script intended to be passed to PSyclone's generate() function
 via the -s option. This script module-inline all kernels in the PSy-layer.'''
```
I think it already said this, or I am misunderstanding what you mean?

sergisiso commented on benchmarks/nemo/tracer_advection/compute_in_subroutine/Makefile:
```diff
@@ -45,7 +45,7 @@ tra_adv_no_auto_serial: dl_timer
 # OpenACC version with timer around outer loop only.
```
Done

sergisiso commented on benchmarks/nemo/tracer_advection/compute_in_subroutine/Makefile:
```diff
@@ -55,7 +55,7 @@ tra_adv_acc: dl_timer
 # OpenACC version with nvtx profiling instrumentation.
```
Done

sergisiso commented on benchmarks/nemo/tracer_advection/compute_in_subroutine/Makefile:
```diff
@@ -45,7 +45,7 @@ tra_adv_no_auto_serial: dl_timer
+	${PSYCLONE} -s ../scripts/acc_kernels_unified_memory_trans.py -opsy \
             $@/tra_adv_compute.f90 ./tra_adv_compute_auto_arrays.F90
 	cp Makefile_gen $@/Makefile
 	cp tra_adv_driver.F90 $@/.
```
I added the flags now. Was this working before? Maybe it had to use the nvidia_acc.sh script which adds the acc flags as part of F90FLAGS, but I want to deprecate this on #94

arporter replied:
Oh! That's probably what it is. I was puzzled that it didn't work.

sergisiso commented on benchmarks/nemo/tracer_advection/compute_in_subroutine/Makefile:
```diff
@@ -55,7 +55,7 @@ tra_adv_acc: dl_timer
+	${PSYCLONE} --profile invokes -s ../scripts/acc_kernels_unified_memory_trans.py -opsy \
             $@/tra_adv_compute.f90 ./tra_adv_compute_auto_arrays.F90
 	cp Makefile_gen $@/Makefile
 	cp tra_adv_driver.F90 $@/.
```
Is is probably because PSYCLONE_NVIDIA_LIB_DIR was not defined. I think this confirms that for this folder the nvidia_acc.sh was used.
I added a proper error reporting for this so at least it is clear the reason for failing.

sergisiso commented on benchmarks/nemo/tracer_advection/multi_kernel/Makefile:
```diff
@@ -36,7 +36,7 @@ tra_adv_serial: dl_timer
-	${PSYCLONE} -s ../scripts/kernels_trans.py -opsy \
+	${PSYCLONE} -s ../scripts/acc_kernels_unified_memory_trans.py -opsy \
             $@/tra_adv_compute.f90 ./tra_adv_compute.F90
 	cp Makefile_gen $@/Makefile
```
I made the same changes as in the compute_is_subroutine folder. Now it should work.

sergisiso commented:
@arporter This is ready for another look. As suggested I split the NemoLite2D part into another PR. The confusion/replication I previously had in the OpenACC mixed versions were because I was trying to conserve the previous code. But it is much nicer using the utils.py (and has better collapsing clauses) so I updated all scripts to use utils.py now.

sergisiso commented:
@arporter Just a reminder that this is ready for another review.

arporter commented on benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_transform.py:
```diff
@@ -15,7 +14,7 @@ def trans(psy):
     loop_trans = tinfo.get_trans_name('GOceanOMPLoopTrans')
     parallel_trans = tinfo.get_trans_name('OMPParallelTrans')
-    module_inline_trans = KernelModuleInlineTrans()
+    module_inline_trans = tinfo.get_trans_name('KernelModuleInline')
```
psyclone.errors.GenerationError: Generation Error: Invalid transformation name: got KernelModuleInline but expected one of dict_keys(['ACCDataTrans', 'ACCEnterDataTrans', 'ACCKernelsTrans', 'ACCLoopTrans', 'ACCParallelTrans', 'ACCRoutineTrans', 'ColourTrans', 'Dynamo0p3AsyncHaloExchangeTrans', 'Dynamo0p3ColourTrans', 'Dynamo0p3KernelConstTrans', 'Dynamo0p3OMPLoopTrans', 'Dynamo0p3RedundantComputationTrans', 'DynamoOMPParallelLoopTrans', 'GOceanOMPLoopTrans', 'GOceanOMPParallelLoopTrans', 'KernelImportsToArguments', 'Move', 'OMPDeclareTargetTrans', 'OMPLoopTrans', 'OMPMasterTrans', 'OMPParallelLoopTrans', 'OMPParallelTrans', 'OMPSingleTrans', 'OMPTaskloopTrans', 'LoopFuseTrans'])

arporter reviewed (changes requested):
I got a bit confused because of the fact that I'd not fully grasped that you'd removed the NemoLite2D aspects of this PR - you may need to ignore some of my comments.
There are just a couple of outstanding comments where you say you've done it but the files don't seem to have been updated.
Once they are fixed then this can go on.

LonelyCat124 commented:
I'm using this branch for CPU OpenMP parallelisation of tracer advection and it hits https://github.com/stfc/PSyclone/issues/2101 as well for me even on this branch - probably not an issue of this branch but worth bearing in mind.

arporter reviewed (approved):
All good now, will proceed to merge.
